### ISSUE TITLE
Help making @dataclass attributes lazy load

### ISSUE THREAD

**Issue by kylepjohnson** [cltk/cltkv1#35]
My problem: I've tried and failed to lazy load the loading of FastText embeddings into memory.

Here's what I am trying to have happen:
* `from cltkv1 import nlp`: The library should load fast, and in this case not load any fasttext models loaded into memory (currently they are all loaded at this step).
* `cltk_nlp = nlp("lat")`: Now that the language has been declared, load into memory only the processes for the Latin language, as declared at [LatinPipeline.processes](https://github.com/cltk/cltkv1/blob/add-embeddings/src/cltkv1/languages/pipelines.py#L144). At this point, [LatinEmbeddingsProcess.algorithm](https://github.com/cltk/cltkv1/blob/add-embeddings/src/cltkv1/embeddings/processes.py#L143) should be lazy-loaded (takes ~60 secs).
* `vergil_doc = cltk_nlp.analyze("arma virumque cano")`: Each process declared in the Latin pipeline is run in succession.
* `livy_doc = cltk_nlp.analyze("iam primum omnium satis constat Troia capta")`: Text is processed as above, only this time the fasttext embedding does not need to be loaded into memory again. I take it this means that each process loaded in [the pipeline loop of NLP()](https://github.com/cltk/cltkv1/blob/add-embeddings/src/cltkv1/nlp.py#L127) is retained inside NLP().

@free-variation to jog your memory, last we talked about this, you helped with this by adding these:
* [Process.run()](https://github.com/cltk/cltkv1/blob/add-embeddings/src/cltkv1/core/data_types.py#L196)
* The closure [make_embedding_algorithm()](https://github.com/cltk/cltkv1/blob/add-embeddings/src/cltkv1/embeddings/processes.py#L14)

@todd-cook this seems like a subject you'd be familiar with, too.

To reproduce my code, pull from the branch `add-embeddings` in this repo: https://github.com/cltk/cltkv1/tree/add-embeddings . Then (with Python 3.7) run the commands in here: https://gist.github.com/kylepjohnson/f88190973fb4b485eba8b111669aae9b .

Extra notes:
* Question: Is lazy loading even possible with dataclasses? In tutorials I see `__getattr__` and a few other patterns recommended, but I have had limited success in applying them (I can go into details if you want).
* Issue comes from #10

**Comment by kylepjohnson** [comment]
Does this help? Adapting a StackOverflow answer recommending `@cachedproperty` decorator from the [boltons library](https://boltons.readthedocs.io/en/latest/cacheutils.html#boltons.cacheutils.cachedproperty).

``` python
from dataclasses import dataclass
import time
from boltons.cacheutils import cachedproperty


@dataclass
class Foo:
    value: int = 4

    @cachedproperty
    def cached_prop(self):
        print("Inside the lazy loaded property. This is only called when `Foo.cached_prop` is explicitly called. Going to sleep for 5 secs ...")
        time.sleep(5)
        self.value += 1
        return self.value


f = Foo()
print(f.value)  # initial value
print(f.cached_prop)  # cached property is calculated
f.value = 1
print(f.cached_prop)  # same value for the cached property - it isn't calculated again
print(f.value)  # the backing value is different (it's essentially unrelated value)
```

which outputs:

``` python
4
Inside the lazy loaded property. This is only called when `Foo.cached_prop` is explicitly called. Going to sleep for 5 secs ...
5
5
1
```

So in our case we change from:

``` python
@dataclass
class LatinEmbeddingsProcess(EmbeddingsProcess):
    algorithm = LATIN_WORD_EMBEDDING
    description: str = "Default embeddings for Latin."
    language: str = "lat"
```

to:

``` python
@dataclass
class LatinEmbeddingsProcess(EmbeddingsProcess):
    description: str = "Default embeddings for Latin."
    language: str = "lat"

   @cachedproperty
   def algorithm(self):
        return self.LATIN_WORD_EMBEDDING
```

**Comment by todd-cook** [comment]
it looks promising but I'm getting some trouble implementing it.
Cached property not callable and a few other hiccups. I'll try a few other
things. Thanks for the pointer.

Todd

**Comment by kylepjohnson** [comment]
Here's an example of what's working for me. I am mid-refactoring it now.

``` python 
@dataclass
class EmbeddingsProcess(Process):
    """To be inherited for each language's embeddings declarations.

    .. note::
        There can be no ``DefaultEmbeddingsProcess`` because word embeddings are naturally language-specific.

    Example: ``EmbeddingsProcess`` <- ``LatinEmbeddingsProcess``

    >>> from cltkv1.core.data_types import Doc
    >>> from cltkv1.embeddings.processes import EmbeddingsProcess
    >>> from cltkv1.core.data_types import Process
    >>> issubclass(EmbeddingsProcess, Process)
    True
    >>> emb_proc = EmbeddingsProcess(input_doc=Doc(raw="some input data"))
    """

    language: str = None

    @cachedproperty
    def algorithm(self):
        return FastTextEmbeddings(iso_code=self.language)

    def run(self):
        tmp_doc = self.input_doc
        embedding_length = None
        embeddings_obj = self.algorithm
        for index, word_obj in enumerate(tmp_doc.words):
            if not embedding_length:
                embedding_length = embeddings_obj.get_embedding_length()
            word_embedding = embeddings_obj.get_word_vector(word=word_obj.string)
            if not isinstance(word_embedding, np.ndarray):
                word_embedding = np.zeros([embedding_length])
            word_obj.embedding = word_embedding
            tmp_doc.words[index] = word_obj
        self.output_doc = tmp_doc
```

**Comment by kylepjohnson** [comment]
Closed by #39
